Add Kubernetes 1.28 release blog article (#42381)
kubernetes/website: content/en/blog/_posts/2023-08-04-kubernetes-1.28-blog.md

**Authors**: [Kubernetes v1.28 Release Team](https://github.com/kubernetes/sig-release/blob/master/releases/release-1.28/release-team.md)

Announcing the release of Kubernetes v1.28 Planternetes, the second release of 2023!

This release consists of 46 enhancements. 20 of those enhancements are entering Alpha, 14 are graduating to Beta, and 12 are graduating to Stable.

## Release Theme And Logo

**Kubernetes v1.28: *Planternetes*

The theme for Kubernetes v1.28 is *Planternetes*.

{{< figure src="/images/blog/2023-08-15-kubernetes-1.28-blog/kubernetes-1.28.png" alt="Kubernetes 1.28 Planternetes logo" class="release-logo" >}}

Each Kubernetes release is the culmination of the hard work of thousands of individuals from our community. The people behind this release are industry veterans, parents, others are students new to the ecosystem, contributing our unique experience to create a collective artifact with global impact.

Much like a garden, our release has ever-changing growth, challenges and opportunities. This theme celebrates the meticulous care, intention and efforts to get the release to where we are today. Harmoniously together, we grow better.

# What's New (Major Themes)

> ## Add Support To Handle Non-graceful Node Shutdown

Reviewer: Is this new, or a graduation to GA?

Author: GA, I have updated it.

When a node is shutdown but not detected by Kubelet's Node Shutdown Manager, the pods that are part of a StatefulSet will be stuck in terminating status on the shutdown node and cannot move to a new running node. This is because Kubelet on the shutdown node is not available to delete the pods so the StatefulSet cannot create a new pod with the same name. If there are volumes used by the pods, the VolumeAttachments will not be deleted from the original shutdown node so the volumes used by these pods cannot be attached to a new running node. As a result, the application running on the StatefulSet cannot function properly. If the original shutdown node comes up, the pods will be deleted by Kubelet and new pods will be created on a different running node. If the original shutdown node does not come up, these pods will be stuck in terminating status on the shutdown node forever.

> This handles node shutdown cases that are not detected by Kubelet's Node Shutdown Manager. The pods will be forcefully deleted in this case, triggering the deletion of the VolumeAttachments, and new pods will be created on a new running node so that application can continue to function.

Reviewer: If this is a graduation to stable, try:
```markdown
## Generally available: cleanup following non-graceful node shutdown

Kubernetes v1.20 introduced cleanup handling following a node shutdown that didn't happen cleanly.
That could be because of a node failure, a forced (hard) reboot, or because of a planned shutdown that
the kubelet's _node shutdown manager_ failed to detect. That can then have an impact on StatefulSets
with a Pod still marked as running on the affected node.

In Kubernetes v1.28, cleanup following a non-graceful node shutdown is generally available, graduating
from beta. Whenever an ungraceful node shutdown occurs, the lack of node heartbeats from the affected
kubelet signals the control plane to delete the affected Pods. That forcible deletion triggers further cleanup
including the deletion of volume attachment record for any storage (PersistentVolumeClaims) that the
Pod was using. The StatefulSet controller can then make a new, replacement Pod on a healthy node.

For more information, read [non-graceful node shutdown](/docs/concepts/architecture/nodes/#non-graceful-node-shutdown) in the Kubernetes documentation.
```
That text should get tech review from SIG Node / SIG Apps, ideally.

Reviewer 2: the SIG-Storage leads the feature.

/cc @xing-yang @msau42

And this feature isn't automatically trriggered by the kubernetes itself, so cluster administators should enable it manually by setting the node with the taint `node.kubernetes.io/out-of-service=out-of-service=nodeshutdown:NoExecute` or `node.kubernetes.io/out-of-service=out-of-service=hardwarefailure:NoExecute`.

A little suggestion changes:

```
If a cluster administator taints the non-recoverable node with `node.kubernetes.io/out-of-service`, that forcible deletion triggers ...
```

Reviewer 3: "cleanup following non-graceful node shutdown" -> We have been calling the feature "non-graceful node shutdown", versus the "graceful node shutdown" feature.
It was introduced as alpha in 1.24.
Yes, it is not automatic. User needs to apply the "out-of-service" taint on the shutdown node.

Reviewer: > "cleanup following non-graceful node shutdown" -> We have been calling the feature "non-graceful node shutdown", versus the "graceful node shutdown" feature.

I know, but I've never liked it. Here's why:
- graceful node shutdown makes the kubelet shutdown more graceful
- “non-graceful node shutdown” **does not** make kubelet shut down less gracefully

The two features are unrelated.

Author: Hey folks, do we need to resolve any debates here? Please advise as I would like to resolve this today. Should I change it to what @sftim suggested for the description? thanks in advance

Author: Hi folks we are going to be consistent with the KEP and keep the non-graceful wording. Thanks all for your help we appreciate it.

    
## Mixed version proxy (alpha) {#mixed-version-proxy}

When a cluster has multiple API servers at mixed versions (such as during an upgrade/downgrade or when runtime-config changes and a rollout happens), not every apiserver can serve every resource at every version.

For Kubernetes v1.28, you can enable the _mixed version proxy_ within the API server's aggregation layer.
The mixed version proxy finds requests that the local API server doesn't recognize but another API server
inside the control plan is able to support. Having found a suitable peer, the aggregation layer proxies
the request to a compatible API server; this is transparent from the client's perspective.
    
When an upgrade or downgrade is performed on a cluster, for some period of time the API servers
within the control plane may be at differing versions; when that happens, different subsets of the
API servers are able to serve different sets of built-in resources (different groups, versions, and resources
are all possible). This new alpha mechanism lets you hide that skew from clients.

## CRD validation ratcheting
    

> This Allows CRs to fail validation if the patch did not alter any of the invalid fields.  The ability to shift left validation logic from controllers to the front end is a long-term goal for improving the useability of the Kubernetes project.

Reviewer: This text needs work, I think.

- “custom resources” not “CRs”
- it's not that Kubernetes allows CRs to fail validation; it's that changes to CRs can now pass admission-time checks, even when there are (certain kinds of) validation failures.

Author: I updated the CR's to custom resources. Do you think it is better if we just take this whole section out of the blog?

Sidecar containers do not block Pod completion: if all regular containers are complete, sidecar
containers in that Pod will be terminated.

During the sidecar startup stage, the restart behavior will be similar to init containers. If the Pod restartPolicy is Never, the sidecar container that failed during startup will NOT be restarted and the whole Pod will fail. If the Pod restartPolicy is Always or OnFailure, it will be restarted.

Once the sidecar container is started (postStart completed and startup probe succeeded), these containers will be restarted even when the Pod restartPolicy is Never or OnFailure. Furthermore, sidecar containers will be restarted even during Pod termination.

In order to minimize OOM kills of sidecar containers, the OOM adjustment for these containers will match or exceed the OOM score adjustment of regular containers in the Pod.

This also enables sidecar containers (those will not be allowed for other init containers):

- PostStart and PreStop lifecycle handlers for sidecar containers
- All probes (startup, readiness, liveness)
    
Readiness probes of sidecars will contribute to determining the whole Pod readiness.

> ## Automatic, retroactive assignment of a default StorageClass

Reviewer 3: Please indicate this is stable in 1.28.

Author: Done, thanks

As a result, the existing implementation did not cover the situation where the workload is truly
[embarrassingly parallel](https://en.wikipedia.org/wiki/Embarrassingly_parallel): each index is
fully independent of other indexes.

For instance, if indexed jobs were used as the basis for a suite of long-running integration tests, then each test run would only be able to find a single test failure.

    
For more information, read [Handling Pod and container failures](/docs/concepts/workloads/controllers/job/#handling-pod-and-container-failures) in the Kubernetes documentation.
   
## CRI container and pod statistics without cAdvisor

This encompasses two related pieces of work (changes to the kubelet's `/metrics/cadvisor` endpoint, and improvements to the replacement _summary_ API).

There are two main APIs that consumers use to gather stats about running containers and pods: summary API and `/metrics/cadvisor`. The Kubelet is responsible for implementing the summary API, and cadvisor is responsible for fulfilling `/metrics/cadvisor`.

This enhances CRI implementations to be able to fulfill all the stats needs of Kubernetes. At a high level, there are two pieces of this:

> - It Enhances the CRI API with enough metrics to be able to supplement the pod and container fields in the summary API directly from CRI.

Reviewer 4: ```suggestion
- It enhances the CRI API with enough metrics to supplement the pod and container fields in the summary API directly from CRI.
```

> - It Enhances the CRI implementations to broadcast the required metrics to fulfill the pod and container fields in the `/metrics/cadvisor` endpoint.

Reviewer 4: ```suggestion
- It enhances the CRI implementations to broadcast the required metrics to fulfill the pod and container fields in the `/metrics/cadvisor` endpoint.
```

> ## Feature Graduations And Deprecations In Kubernetes v1.28

Reviewer 4: ```suggestion
## Feature graduations and deprecations in Kubernetes v1.28
```
Starting from this line, the headings below use a different capitalization style compared to the ones above.

* [Non-graceful node shutdown](https://github.com/kubernetes/enhancements/issues/2268)
* [Support 3rd party device monitoring plugins](https://github.com/kubernetes/enhancements/issues/606)
* [Auth API to get self-user attributes](https://github.com/kubernetes/enhancements/issues/3325)
* [Proxy Terminating Endpoints](https://github.com/kubernetes/enhancements/issues/1669)
* [Expanded DNS Configuration](https://github.com/kubernetes/enhancements/issues/2595)
* [Cleaning up IPTables Chain Ownership](https://github.com/kubernetes/enhancements/issues/3178)
* [Minimizing iptables-restore input size](https://github.com/kubernetes/enhancements/issues/3453)
* [Graduate the kubelet pod resources endpoint to GA](https://github.com/kubernetes/enhancements/issues/3743)
* [Extend podresources API to report allocatable resources](https://github.com/kubernetes/enhancements/issues/2403)
* [Move EndpointSlice Reconciler into Staging](https://github.com/kubernetes/enhancements/issues/3685)

### Deprecations And Removals

This release saw one removal:

* [Removal of CSI Migration for GCE PD](https://github.com/kubernetes/enhancements/issues/1488)

> 

Reviewer: Any deprecations?

Reviewer 3: Deprecations:
Ceph RBD in-tree plugin:
https://github.com/kubernetes/kubernetes/pull/118303
Ceph FS in-tree plugin:
https://github.com/kubernetes/enhancements/pull/4035

Reviewer 2: https://github.com/kubernetes/kubernetes/pull/118143 deprecate CephFS plugin from available in-tree drivers

Author: Added thanks.

---
layout: blog
title: "Kubernetes v1.28: Planternetes"
date: 2023-08-15T12:00:00+0000
slug: kubernetes-v1-28-release
---

**Authors**: [Kubernetes v1.28 Release Team](https://github.com/kubernetes/sig-release/blob/master/releases/release-1.28/release-team.md)

Announcing the release of Kubernetes v1.28 Planternetes, the second release of 2023!

> This release consists of 46 enhancements. 20 of those enhancements are entering Alpha, 14 are graduating to Beta, and 12 are graduating to Stable.

Reviewer: (nit)
```suggestion
This release consists of 46 enhancements. 20 of those enhancements are entering Alpha, 14 have graduated to Beta, and 12 have graduated to Stable.
```
?

When a node is shutdown but not detected by Kubelet's Node Shutdown Manager, the pods that are part of a StatefulSet will be stuck in terminating status on the shutdown node and cannot move to a new running node. This is because Kubelet on the shutdown node is not available to delete the pods so the StatefulSet cannot create a new pod with the same name. If there are volumes used by the pods, the VolumeAttachments will not be deleted from the original shutdown node so the volumes used by these pods cannot be attached to a new running node. As a result, the application running on the StatefulSet cannot function properly. If the original shutdown node comes up, the pods will be deleted by Kubelet and new pods will be created on a different running node. If the original shutdown node does not come up, these pods will be stuck in terminating status on the shutdown node forever.

This handles node shutdown cases that are not detected by Kubelet's Node Shutdown Manager. The pods will be forcefully deleted in this case, triggering the deletion of the VolumeAttachments, and new pods will be created on a new running node so that application can continue to function.
    

> ## Improvements to CustomResourceDefinition validation expressions 

Reviewer 5: We added a `reason` field to allow CRD authors to provide more more context to validation failures. We also added a `fieldPath` field to allow the validation error to be attributed to a specific field, which is useful in many cross field validation rules.

This [Common Expression Language (CEL)](https://github.com/google/cel-go) can be used to validate custom resources. The primary goal is to allow the majority of the validation use cases that currently must be handled by a webhook, to instead be handled by adding inline validation expressions directly into the schema of a CRD.
    

> CRDs need direct support for non-trivial validation. While admission webhooks do support CRDs validation, they significantly complicate the development and operability of CRDs.

Author: We are waiting to hear back from the sig-api-machinery to help out here, as we tried to look today for this but couldnt see/understand.

It is sufficiently lightweight and safe to be run directly in the kube-apiserver (since CRD creation is a privileged operation), has a straight-forward and unsurprising grammar, and supports pre-parsing and type checking of expressions, allowing syntax and type errors to be caught at CRD registration time.

## ValidatingAdmissionPolicies graduate to beta
    
Common Expression language for admission control is customizable, in-process validation of requests to the Kubernetes API server as an alternative to validating admission webhooks.

This builds on the capabilities of the CRD Validation Rules feature that graduated to beta in 1.25 but with a focus on the policy enforcement capabilities of validating admission control.
    
This will lower the infrastructure barrier to enforcing customizable policies as well as providing primitives that help the community establish and adhere to the best practices of both K8s and its extensions.

## CRD validation ratcheting
    

> This Allows custom resources to fail validation if the patch did not alter any of the invalid fields.  The ability to shift left validation logic from controllers to the front end is a long-term goal for improving the useability of the Kubernetes project.

Reviewer: I'd move _Improvements to CustomResourceDefinition validation rules_ so that it is after the ValidatingAdmissionPolicies item and immediately before this one.

  
## Match conditions for admission webhooks
    
Kubernetes v1.27 lets you specify _match conditions_ for admission webhooks,
which lets you narrow the scope of when Kubernetes makes a remote HTTP call at admission time.
The `matchCondition` field for ValidatingWebhookConfiguration and MutatingWebhookConfiguration
is a CEL expression that must evaluate to true for the admission request to be sent to the webhook.

In Kubernetes v1.28, that field moved to beta, and it's enabled by default.
    
To learn more, see [`matchConditions`](/docs/reference/access-authn-authz/extensible-admission-controllers/#matching-requests-matchconditions) in the Kubernetes documentation.

## Changes to supported skew between control plane and node versions
    

> This enables testing and expanding the supported skew between core node and control plane components by one version from n-2 to n-3, so that node components (kubelet and kube-proxy) for the oldest supported minor version work with control plane components (kube-apiserver, kube-scheduler, kube-controller-manager, cloud-controller-manager) for the newest supported minor version.

Reviewer 6: I think this is important for end users. It essentially means they can upgrade the control plane a little faster than their nodes (which are almost always going to be the longer thing with running workloads). 

If I was on 1.25 and I wanted to go to 1.28 previously, I would need to upgrade my control plane to 1.26, then 1.27, then I would need to upgrade my nodes to 1.26, then I could upgrade my control plane to 1.28. This will allow upgrades of the control plan 1.25 -> 1.26 -> 1.27 -> 1.28 and then I could start upgrading my nodes.

> This enables testing and expanding the supported skew between core node and control plane components by one version from n-2 to n-3, so that node components (kubelet and kube-proxy) for the oldest supported minor version work with control plane components (kube-apiserver, kube-scheduler, kube-controller-manager, cloud-controller-manager) for the newest supported minor version.

Reviewer 6: TL;DR I think this makes it better align with the yearly support period and reduces operational burden on kubernetes operators.

Once the sidecar container is started (postStart completed and startup probe succeeded), these containers will be restarted even when the Pod restartPolicy is Never or OnFailure. Furthermore, sidecar containers will be restarted even during Pod termination.

In order to minimize OOM kills of sidecar containers, the OOM adjustment for these containers will match or exceed the OOM score adjustment of regular containers in the Pod.

This also enables sidecar containers (those will not be allowed for other init containers):

- PostStart and PreStop lifecycle handlers for sidecar containers
- All probes (startup, readiness, liveness)
    
Readiness probes of sidecars will contribute to determining the whole Pod readiness.

## Automatic, retroactive assignment of a default StorageClass graduates to stable
    
This feature makes it easier to change the default StorageClass by allowing the default storage class assignment to be retroactive for existing unbound persistent volume claims without any storage class assigned.
    

> This changes the behavior of default storage class assignment to be retroactive for existing unbound persistent volume claims without any storage class assigned. This changes the existing Kubernetes behavior slightly, which is further described in the sections below.

Reviewer: - This should move up before the beta items
- What sections below?

* [Support 3rd party device monitoring plugins](https://github.com/kubernetes/enhancements/issues/606)
* [Auth API to get self-user attributes](https://github.com/kubernetes/enhancements/issues/3325)
* [Proxy Terminating Endpoints](https://github.com/kubernetes/enhancements/issues/1669)
* [Expanded DNS Configuration](https://github.com/kubernetes/enhancements/issues/2595)
* [Cleaning up IPTables Chain Ownership](https://github.com/kubernetes/enhancements/issues/3178)
* [Minimizing iptables-restore input size](https://github.com/kubernetes/enhancements/issues/3453)
* [Graduate the kubelet pod resources endpoint to GA](https://github.com/kubernetes/enhancements/issues/3743)
* [Extend podresources API to report allocatable resources](https://github.com/kubernetes/enhancements/issues/2403)
* [Move EndpointSlice Reconciler into Staging](https://github.com/kubernetes/enhancements/issues/3685)

### Deprecations And Removals

Removals:

* [Removal of CSI Migration for GCE PD](https://github.com/kubernetes/enhancements/issues/1488)

Deprecations:

* [Ceph RBD in-tree plugin](https://github.com/kubernetes/kubernetes/pull/118303)

> * [Ceph FS in-tree plugin](https://github.com/kubernetes/enhancements/pull/4035)

Reviewer 2: https://github.com/kubernetes/kubernetes/pull/118143 is right link
```suggestion
* [Ceph FS in-tree plugin](https://github.com/kubernetes/kubernetes/pull/118143)
```

Reviewer: We could expand “FS”, too.

---
layout: blog
title: "Kubernetes v1.28: <name-here>"
date: 2023-08-02
slug: kubernetes-v1-28-release
---

**Authors**: [Kubernetes v1.28 Release Team](https://github.com/kubernetes/sig-release/blob/master/releases/release-1.28/release-team.md)

Announcing the release of Kubernetes v1.28 <name-here>, the second release of 2023!

This release consists of 46 enhancements. 20 of those enhancements are entering Alpha, 14 are graduating to Beta, and 12 are graduating to Stable.

## Release Theme And Logo

**Kubernetes v1.28: *<name-here>*

The theme for Kubernetes v1.28 is *<name-here>*.

{{< figure src="/images/blog/2023-07-02-kubernetes-1.28-blog/kubernetes-1.28.png" alt="Kubernetes 1.28 <name-here> logo" class="release-logo" >}}

Talk about release <here> - waiting for the theme

Special thanks to [someone](https://www.some-link) for creating the logo.

# What's New (Major Themes)
    

> ## CRD Validation Rules

Reviewer: ```suggestion
## Improvements to CustomResourceDefinition validation expressions 
```

This [Common Expression Language (CEL)](https://github.com/google/cel-go) can be used to validate custom resources. The primary goal is to allow the majority of the validation use cases that currently must be handled by a webhook, to instead be handled by adding inline validation expressions directly into the schema of a CRD.
    
CRDs need direct support for non-trivial validation. While admission webhooks do support CRDs validation, they significantly complicate the development and operability of CRDs.

This KEP proposes that an inline expression language be integrated directly into CRDs such that a much larger portion of validation use cases can be solved without the use of webhooks. When selecting an expression language, we want to be sure that it can support defaulting and CRD conversion in the future.

It is sufficiently lightweight and safe to be run directly in the kube-apiserver (since CRD creation is a privileged operation), has a straight-forward and unsurprising grammar, and supports pre-parsing and type checking of expressions, allowing syntax and type errors to be caught at CRD registration time.

> ## Common Expression Language For Admission Control

Reviewer: ```suggestion
## ValidatingAdmissionPolicies graduate to beta
```

    
Common Expression language for admission control is customizable, in-process validation of requests to the Kubernetes API server as an alternative to validating admission webhooks.

This builds on the capabilities of the CRD Validation Rules feature that graduated to beta in 1.25 but with a focus on the policy enforcement capabilities of validating admission control.
    
This will lower the infrastructure barrier to enforcing customizable policies as well as providing primitives that help the community establish and adhere to the best practices of both K8s and its extensions.
        

> ## Admission Webhook Match Conditions

Reviewer: ```suggestion
## Match conditions for admission webhooks
```

    
Introduce CEL expression filters to webhooks, to allow webhooks to be scoped more narrowly.

This adds "match conditions" to admission webhooks, as an extension to the existing rules to define the scope of a webhook. A matchCondition is a CEL expression that must evaluate to true for the admission request to be sent to the webhook. If a matchCondition evaluates to false, the webhook is skipped for that request (implicitly allowed).
    
ValidatingAdmissionPolicy is an exciting new feature that we hope will greatly reduce the need for admission webhooks, but it is intentionally not attempting to cover every possible use case. This proposal aims to improve the situation for those webhooks that cannot be migrated.
    

> ## Unknown Version Interoperability Proxy

Reviewer: According to https://github.com/kubernetes/website/pull/42186, this should be:
```suggestion
## Mixed version proxy
```

When a cluster has multiple API servers at mixed versions (such as during an upgrade/downgrade or when runtime-config changes and a rollout happens), not every apiserver can serve every resource at every version.

To fix this, the filter is added to the handler chain in the aggregator which proxies clients to an API server that is capable of handling their request.
    
When an upgrade or downgrade is performed on a cluster, for some period of time the api servers are at differing versions and are able to serve different sets of built-in resources (different groups, versions, and resources are all possible).

## CRD Validation Ratcheting
    

> This Allows CRs to fail validation if the patch did not alter any of the invalid fields.  The ability to shift left validation logic from controllers to the front end is a long-term goal for improving the useability of the Kubernetes project.

Reviewer: I'd list this along with the other change to CRD validation.

    
## Add Generic Control Plane Staging Repositories

This factors the kube-apiserver to build on a new staging repository that consumes k/apiserver but has a bigger, carefully chosen subset of the functionality of kube-apiserver such that it is reusable.

The factoring will be progressive: we will start with a new repo that adds
nothing to k/apiserver, and then progressively move generic functionality from
kube-apiserver to the new repository. The new repo will be named
k/generic-controlplane.
  
## CDI Support To The Device Plugin API

CDI provides a standardized way of injecting complex devices into a container (i.e. devices that logically require more than just a single /dev node to be injected for them to work). This new feature enables plugin developers to utilize the CDIDevices field added to the CRI in 1.27 to pass CDI devices directly to CDI enabled runtimes (of which containerd and crio-o are in recent releases).
    

> ## Sidecar Containers Built-in Support

Reviewer: ```suggestion
## API awareness of sidecar containers (alpha)
```

> This introduces a restartPolicy field to init containers and uses it to indicate that an init container is a sidecar container. Kubelet will start init containers with restartPolicy=Always in the order with other init containers, but instead of waiting for its completion, it will wait for the container startup completion.

Reviewer: ```suggestion
Kubernetes 1.28 introduces an alpha `restartPolicy` field for [init containers](/docs/concepts/workloads/pods/init-containers/),
and uses that to indicate when an init container is also a _sidecar container_. The will start init containers with `restartPolicy: Always` in the order they are defined, along with other init containers. Instead of waiting for that sidecar container to complete before starting the main container(s) for the Pod, the kubelet only waits for
the sidecar init container to have started.
```

> The condition for startup completion will be that the startup probe succeeded (or if no startup probe is defined) and postStart handler is completed. This condition is represented with the field Started of ContainerStatus type. See the section "Pod startup completed condition" for considerations on picking this signal.

Reviewer: We should add a hyperlink

Author: Thanks done.

> The field restartPolicy will only be accepted on init containers. The only supported value now is Always. No other values will be defined. Moreover, the field will be nullable so the default value will be "no value".
> 
> Other values for restartPolicy of containers will not be accepted and containers will follow the logic currently implemented.

Reviewer: ```suggestion
For init containers, you can either omit the `restartPolicy` field, or set it to `Always`. Omitting the field
means that you want a true init container that runs to completion before application startup.
```

> Sidecar containers will not block Pod completion - if all regular containers are complete, sidecar containers will be terminated.

Reviewer: ```suggestion
Sidecar containers do not block Pod completion: if all regular containers are complete, sidecar
containers in that Pod will be terminated.
```

During the sidecar startup stage, the restart behavior will be similar to init containers. If the Pod restartPolicy is Never, the sidecar container that failed during startup will NOT be restarted and the whole Pod will fail. If the Pod restartPolicy is Always or OnFailure, it will be restarted.

Once the sidecar container is started (postStart completed and startup probe succeeded), these containers will be restarted even when the Pod restartPolicy is Never or OnFailure. Furthermore, sidecar containers will be restarted even during Pod termination.

In order to minimize OOM kills of sidecar containers, the OOM adjustment for these containers will match or exceed the OOM score adjustment of regular containers in the Pod.

This also enables sidecar containers (those will not be allowed for other init containers):

- PostStart and PreStop lifecycle handlers for sidecar containers
- All probes (startup, readiness, liveness)
    

> Readiness probes of sidecars will contribute to determining the whole Pod readiness.

Reviewer: We should just link to the docs (once they exist) rather than give more detail here.

Reviewer: > We should just link to the docs (once they exist) rather than give more detail here.

What's happening on this?

Reviewer 7: We can point to this: https://github.com/kubernetes/website/blob/dev-1.28/content/en/docs/concepts/workloads/pods/init-containers.md

Author: Thanks done.

> ## Node System Memory Swap Support

Reviewer: This should be **top** of the major themes.

```suggestion
## Beta support for enabling swap space on Linux
```

    
This adds swap support to nodes in a controlled, predictable manner so that Kubernetes users can perform testing and provide data to continue building cluster capabilities on top of swap.

There are two distinct types of users for swap, who may overlap:

- Node administrators, who may want swap available for node-level performance tuning and stability/reducing noisy neighbor issues.
    
- Application developers, who have written applications that would benefit from using swap memory. There are hence a number of possible ways that one could envision swap use on a node.
   

> ## Add Support To Handle Non-graceful Node Shutdown

Reviewer: Is this actually a graduation to GA?

Reviewer 2: It‘s a GA feature in 1.28 @bradmccoydev

When a node is shutdown but not detected by Kubelet's Node Shutdown Manager, the pods that are part of a StatefulSet will be stuck in terminating status on the shutdown node and cannot move to a new running node. This is because Kubelet on the shutdown node is not available to delete the pods so the StatefulSet cannot create a new pod with the same name. If there are volumes used by the pods, the VolumeAttachments will not be deleted from the original shutdown node so the volumes used by these pods cannot be attached to a new running node. As a result, the application running on the StatefulSet cannot function properly. If the original shutdown node comes up, the pods will be deleted by Kubelet and new pods will be created on a different running node. If the original shutdown node does not come up, these pods will be stuck in terminating status on the shutdown node forever.

This handles node shutdown cases that are not detected by Kubelet's Node Shutdown Manager. The pods will be forcefully deleted in this case, triggering the deletion of the VolumeAttachments, and new pods will be created on a new running node so that application can continue to function.

> ## Retroactive Default StorageClass Assignment

Reviewer: Is this actually a graduation to GA?
```suggestion
## Automatic, retroactive assignment of a default StorageClass
```

When users want to change the default SC parameters, they must delete the SC and re-create it, Kubernetes API does not allow change in the SC. So there is no default SC for some time and the second case above applies here too. Re-creating the storage class to change parameters can be useful in cases where there is a quota set for the SC, and since the quota is coupled with the SC name users can not use Option 1 because the second SC would have a different name and so the existing quota would not apply to it.

Defined ordering during cluster installation. Kubernetes cluster installation tools must be currently smart enough to create a default SC before starting anything that may create PVCs that need it. If such a tool supports multiple cloud providers, storage backends, and add-ons that require storage (such as an image registry), it may be quite complicated to do the ordering right.
          

> ## Support The Oldest Node And Newest Control Plane

Reviewer: I'd have this as the second major theme; it's an important change.
```suggestion
## Changes to supported skew between control plane and node versions
```

There can be many more nodes (hundreds to thousands) than control plane members (typically 1 or 3).
    
Every time nodes are upgraded to a new minor version, every pod running on those nodes must be drained/rescheduled. This is true for immutable nodes and mutable/bare-metal nodes. If all nodes are being upgraded, this means every pod in the cluster will be replaced at least once. Patch updates of kubelet / kube-proxy components can be done in place, so it is possible to pick up security fixes and patch updates less disruptively.
    
Replacing or moving pods that are slow to stop or start or have significant data gravity takes significant time, so it is desirable to minimize how frequently that must be done.
If node/control plane skew support was expanded so the oldest node components work with the newest control plane components, the example upgrade path from v1.40 to v1.43 above could improve this:

Begin: control plane and nodes on v1.40
Control plane upgrade: v1.40 → v1.41 → v1.42 → v1.43
Node upgrades: v1.40 → v1.43
   

> ## Pod Replacement Policy

Reviewer: ```suggestion
## Pod replacement policy for Jobs
```

    
Many common machine learning frameworks, such as Tensorflow and JAX, require unique pods per Index. Currently, if a pod enters a terminating state (due to preemption, eviction or other external factors), a replacement pod is created and immediately fails to start.

Having a replacement Pod before the previous one fully terminates can also cause problems in clusters with scarce resources or with tight budgets. These resources can be difficult to obtain so pods can take a long time to find resources and they may only be able to find nodes once the existing pods have been terminated. If cluster autoscaler is enabled, the replacement Pods might produce undesired scale-ups.

On the other hand, if a replacement Pod is not immediately created, the Job status would show that the number of active pods doesn't match the desired parallelism. To provide better visibility, the job status can have a new field to track the number of Pods currently terminating.

This new field can also be used by queueing controllers, such as Kueue, to track the number of terminating pods to calculate quotas.
  

> ## Backoff Limit Per Index

Reviewer: We could combine this and the previous item. If not:
```suggestion
## Job retry backoff limit, per index
```

As a result, the current implementation does not cover the situation where the workload is truly embarrassingly parallel and each index is independent of other indexes.

For instance, if indexed jobs were used as the basis for a suite of long-running integration tests, then each test run would only be able to find a single test failure.

Other popular batch services like AWS Batch use a separate backoff limit for each index, showing that this is a common use case that should be supported by Kubernetes.
    
We propose a new policy for running Indexed Jobs in which the backoff limit controls the number of retries per index. When the new policy is used all indexes execute until their success or failure. We also propose a new API field to control the number of failed indexes.

Additionally, we propose a new action in PodFailurePolicy, called FailIndex, to short-circuit the failing of the index before the backoff limit per index is reached.
   
## cAdvisor-less, CRI-full Container And Pod Stats

> This encompasses two related pieces of work (summary API and /metrics/cadvisor), and will require changes in three different components (CRI implementation, Kubelet, cAdvisor).

Reviewer: ```suggestion
This encompasses two related pieces of work (changes to the kubelet's `/metrics/cadvisor` endpoint, and improvements to the replacement _summary_ API).
```
Don't imply that readers need to make changes in three places unless that is what we do want to tell people.

There are two main APIs that consumers use to gather stats about running containers and pods: summary API and /metrics/cadvisor. The Kubelet is responsible for implementing the summary API, and cadvisor is responsible for fulfilling /metrics/cadvisor.

This aims to enhance CRI implementations to be able to fulfill all the stats needs of Kubernetes. At a high level, there are two pieces of this:

- Enhance the CRI API with enough metrics to be able to supplement the pod and container fields in the summary API directly from CRI.
    
- Enhance the CRI implementations to broadcast the required metrics to fulfill the pod and container fields in the /metrics/cadvisor endpoint.

## Feature Graduations And Deprecations In Kubernetes v1.28
### Graduations to stable

This release includes a total of 12 enhancements promoted to Stable:

> * [Kubectl Events](https://github.com/kubernetes/enhancements/issues/1440)

Reviewer: ```suggestion
* [`kubectl events`](https://github.com/kubernetes/enhancements/issues/1440)
```

* [Retroactive default StorageClass assignment](https://github.com/kubernetes/enhancements/issues/3333)
* [Non-graceful node shutdown](https://github.com/kubernetes/enhancements/issues/2268)
* [Support 3rd party device monitoring plugins](https://github.com/kubernetes/enhancements/issues/606)
* [Auth API to get self-user attributes](https://github.com/kubernetes/enhancements/issues/3325)
* [Proxy Terminating Endpoints](https://github.com/kubernetes/enhancements/issues/1669)
* [Expanded DNS Configuration](https://github.com/kubernetes/enhancements/issues/2595)
* [Cleaning up IPTables Chain Ownership](https://github.com/kubernetes/enhancements/issues/3178)
* [Minimizing iptables-restore input size](https://github.com/kubernetes/enhancements/issues/3453)
* [Graduate the kubelet pod resources endpoint to GA](https://github.com/kubernetes/enhancements/issues/3743)
* [Extend podresources API to report allocatable resources](https://github.com/kubernetes/enhancements/issues/2403)
* [Move EndpointSlice Reconciler into Staging](https://github.com/kubernetes/enhancements/issues/3685)

### Deprecations And Removals

This release saw several removals:

> * [Removal of CSI Migration for GCE PD](https://github.com/kubernetes/enhancements/issues/1488)

Reviewer 8: Note: there is only single removal based on the current state of the PR, and we might need to reword line 236

Author: Thanks @furkatgofurov7 I was wondering about this. What do you recommend I reword it to??

Reviewer 2: Deprecations of Volume Plugin: 

[Ceph RBD in-tree Plugin](https://groups.google.com/g/kubernetes-sig-storage/c/h5751_B5LQM) has been deprecated in v1.28 and is targeted for removal in v1.31 (no CSI migration planned).

[Ceph FS in-tree Plugin](https://groups.google.com/g/kubernetes-sig-storage/c/IlHnIpwTbEg) has been deprecated in v1.28 and is targeted for removal in v1.31 (no CSI migration planned).

PRs:https://github.com/kubernetes/website/pull/42201

Reviewer 8: @bradmccoydev perhaps it can stay as is if we have multiple removals based on @carlory s comments, otherwise I was thinking of something like:

`This release saw only one removal:`

could be a better fit.

 
## Release Team

Kubernetes is only possible with the support, commitment, and hard work of its community. Each release team is made up of dedicated community volunteers who work together to build the many pieces that make up the Kubernetes releases you rely on. This requires the specialized skills of people from all corners of our community, from the code itself to its documentation and project management.

We would like to thank the entire release team for the hours spent hard at work to ensure we deliver a solid Kubernetes v1.28 release for our community.

Special thanks to our Release Lead, Grace Nguyen, for guiding us through a smooth and successful release cycle and facilitating the communication and collaboration between the different Release Teams taking advantage of her experience as a Shadow and Lead in these teams.

## Ecosystem Updates

> * KubeCon + CloudNativeCon North America 2023 will take place in Chicago, Illinois, The United States of America, from 6 – 9 November 2023! You can find more information about the conference and registration on the [event site](https://events.linuxfoundation.org/kubecon-cloudnativecon-north-america/).

Reviewer 9: Could we mention that Kubecon China will be held on 26 – 28 SEPTEMBER, Shanghai?
-  https://www.lfasiallc.com/kubecon-cloudnativecon-open-source-summit-china/

Author: Thanks totally forgot about that. I have added it.

It is sufficiently lightweight and safe to be run directly in the kube-apiserver (since CRD creation is a privileged operation), has a straight-forward and unsurprising grammar, and supports pre-parsing and type checking of expressions, allowing syntax and type errors to be caught at CRD registration time.

## ValidatingAdmissionPolicies graduate to beta
    
Common Expression language for admission control is customizable, in-process validation of requests to the Kubernetes API server as an alternative to validating admission webhooks.

This builds on the capabilities of the CRD Validation Rules feature that graduated to beta in 1.25 but with a focus on the policy enforcement capabilities of validating admission control.
    
This will lower the infrastructure barrier to enforcing customizable policies as well as providing primitives that help the community establish and adhere to the best practices of both K8s and its extensions.
        
## Match conditions for admission webhooks
    

> Introduce CEL expression filters to webhooks, to allow webhooks to be scoped more narrowly.
> 
> This adds "match conditions" to admission webhooks, as an extension to the existing rules to define the scope of a webhook. A matchCondition is a CEL expression that must evaluate to true for the admission request to be sent to the webhook. If a matchCondition evaluates to false, the webhook is skipped for that request (implicitly allowed).

Reviewer: ```suggestion
Kubernetes v1.27 lets you specify _match conditions_ for admission webhooks,
which lets you narrow the scope of when Kubernetes makes a remote HTTP call at admission time.
The `matchCondition` field for ValidatingWebhookConfiguration and MutatingWebhookConfiguration
is a CEL expression that must evaluate to true for the admission request to be sent to the webhook.

In Kubernetes v1.28, that field moved to beta, and it's enabled by default.
```

> ValidatingAdmissionPolicy is an exciting new feature that we hope will greatly reduce the need for admission webhooks, but it is intentionally not attempting to cover every possible use case. This proposal aims to improve the situation for those webhooks that cannot be migrated.

Reviewer: ```suggestion
To learn more, see [`matchConditions`](/docs/reference/access-authn-authz/extensible-admission-controllers/#matching-requests-matchconditions) in the Kubernetes documentation.
```

> ## Mixed version proxy

Reviewer: ```suggestion
## Mixed version proxy (alpha) {#mixed-version-proxy}
```

When a cluster has multiple API servers at mixed versions (such as during an upgrade/downgrade or when runtime-config changes and a rollout happens), not every apiserver can serve every resource at every version.

> To fix this, the filter is added to the handler chain in the aggregator which proxies clients to an API server that is capable of handling their request.

Reviewer: ```suggestion
For Kubernetes v1.28, you can enable the _mixed version proxy_ within the API server's aggregation layer.
The mixed version proxy finds requests that the local API server doesn't recognize but another API server
inside the control plan is able to support. Having found a suitable peer, the aggregation layer proxies
the request to a compatible API server; this is transparent from the client's perspective.
```

> When an upgrade or downgrade is performed on a cluster, for some period of time the api servers are at differing versions and are able to serve different sets of built-in resources (different groups, versions, and resources are all possible).

Reviewer: ```suggestion
When an upgrade or downgrade is performed on a cluster, for some period of time the API servers
within the control plane may be at differing versions; when that happens, different subsets of the
API servers are able to serve different sets of built-in resources (different groups, versions, and resources
are all possible). This new alpha mechanism lets you hide that skew from clients.
```
?

---
layout: blog
title: "Kubernetes v1.28: Planternetes"
date: 2023-08-15T12:00:00+0000
slug: kubernetes-v1-28-release
---

**Authors**: [Kubernetes v1.28 Release Team](https://github.com/kubernetes/sig-release/blob/master/releases/release-1.28/release-team.md)

Announcing the release of Kubernetes v1.28 Planternetes, the second release of 2023!

> This release consists of 46 enhancements. 20 of those enhancements are entering Alpha, 14 have graduated to Beta, and 12 have graduated to Stable.

Reviewer 10: We got 45 Enhancements, right? 

The KEP https://github.com/kubernetes/enhancements/issues/4138 is a repost from https://github.com/kubernetes/enhancements/issues/3085 and both appear on the project board https://github.com/orgs/kubernetes/projects/140/views/1

Author: Thanks, done.

> This release consists of 46 enhancements. 20 of those enhancements are entering Alpha, 14 have graduated to Beta, and 12 have graduated to Stable.

Reviewer 11: ```suggestion
This release consists of 46 enhancements. Of those enhancements, 20 are entering Alpha, 14 have graduated to Beta, and 12 have graduated to Stable.
```

## Release Theme And Logo

**Kubernetes v1.28: _Planternetes_**

The theme for Kubernetes v1.28 is *Planternetes*.

{{< figure src="/images/blog/2023-08-15-kubernetes-1.28-blog/kubernetes-1.28.png" alt="Kubernetes 1.28 Planternetes logo" class="release-logo" >}}

> Each Kubernetes release is the culmination of the hard work of thousands of individuals from our community. The people behind this release are industry veterans, parents, others are students new to the ecosystem, contributing our unique experience to create a collective artifact with global impact.

Reviewer 12: Could we make this sentence a little more comprehensive to include folks that don't fall into those three buckets?

Author: Thanks we have updated it.

Much like a garden, our release has ever-changing growth, challenges and opportunities. This theme celebrates the meticulous care, intention and efforts to get the release to where we are today. Harmoniously together, we grow better.

# What's New (Major Themes)

## Generally available: recovery from non-graceful node shutdown
    

> If a node is shut down unexpectedly or ends up in a non-recoverable state (perhaps due to hardware failure or unresponsive OS), Kubernetes allows you to clean up afterards and allow stateful workloads to restart on a different node. For Kubernetes v1.28, that's now a stable feature.

Reviewer 7: there is a typo in `afterwards`

Reviewer: ```suggestion
If a node is shut down unexpectedly or ends up in a non-recoverable state (perhaps due to hardware failure or unresponsive OS), Kubernetes allows you to clean up afterwards and allow stateful workloads to restart on a different node. For Kubernetes v1.28, that's now a stable feature.
```

> This allows stateful workloads to failover to a different node successfully after the original node is shut down or in a non-recoverable state such as the hardware failure or broken OS.

Reviewer 13: nit:
```suggestion
This allows stateful workloads to failover to a different node successfully after the original node is shut down or in a non-recoverable state, such as the hardware failure or broken OS.
```

> Kubernetes before 1.20 lacked handling for node shutdown; on Linux, the kubelet integrates with systemd

Reviewer 11: ```suggestion
Versions of Kubernetes earlier than v1.20 lacked handling for node shutdown; on Linux, the kubelet integrates with systemd
```

> Kubernetes before 1.20 lacked handling for node shutdown; on Linux, the kubelet integrates with systemd

Reviewer: @mickeyboxell if you've got the capacity to, use a GitHub suggestion to propose new text

> Kubernetes before 1.20 lacked handling for node shutdown; on Linux, the kubelet integrates with systemd

Author: Thanks done.

and implements _graceful node shutdown_ (beta, and enabled by default). However, even an intentional
shutdown might not get handled well; that could be because:

- the node runs Windows

> - the node runs Linux, but uses a different `init` (not systemd)

Reviewer 13: nit:
```suggestion
- the node runs Linux, but uses a different `init` (not `systemd`)
```

- the shutdown does not trigger the system inhibitor locks mechanism
- because of a node-level configuration error
  (such as not setting appropriate values for `shutdownGracePeriod` and `shutdownGracePeriodCriticalPods`).

> When a node is shutdown or fails, and that shutdown was not detected by the kubelet, the pods that are part

Reviewer 13: ```suggestion
When a node shutdowns or fails, and that shutdown was not detected by the kubelet, the pods that are part
```

> When a node is shutdown or fails, and that shutdown was not detected by the kubelet, the pods that are part

Reviewer 12: ```suggestion
When a node shuts down or fails, and that shutdown was not detected by the kubelet, the pods that are part
```

Author: Thanks done.

of a StatefulSet will be stuck in terminating status on the shutdown node. If the stopped node restarts, the
kubelet on that node can clean up (`DELETE`) the Pods that the Kubernetes API still sees as bound to that node.
However, if the node stays stopped - or if the kubelet isn't able to start after a reboot - then Kubernetes may
not be able to create replacement Pods. When the kubelet on the shut-down node is not available to delete
the old pods, an associated StatefulSet cannot create a new pod (which would have the same name).

There's also a problem with storage. If there are volumes used by the pods, existing VolumeAttachments will
not be disassociated from the original - and now shut down - node so the PersistentVolumes used by these
pods cannot be attached to a different, healthy node. As a result, an application running on an

> affected StatefulSet may not be able to function properly. If the original, shut down node does come up, the

Reviewer 13: ```suggestion
affected StatefulSet may not be able to function properly. If the original, shut down node does come up, then
```

This KEP proposes that an inline expression language be integrated directly into CRDs such that a much larger portion of validation use cases can be solved without the use of webhooks. When selecting an expression language, we want to be sure that it can support defaulting and CRD conversion in the future.

It is sufficiently lightweight and safe to be run directly in the kube-apiserver (since CRD creation is a privileged operation), has a straight-forward and unsurprising grammar, and supports pre-parsing and type checking of expressions, allowing syntax and type errors to be caught at CRD registration time.

## ValidatingAdmissionPolicies graduate to beta
    
Common Expression language for admission control is customizable, in-process validation of requests to the Kubernetes API server as an alternative to validating admission webhooks.

This builds on the capabilities of the CRD Validation Rules feature that graduated to beta in 1.25 but with a focus on the policy enforcement capabilities of validating admission control.
    

> This will lower the infrastructure barrier to enforcing customizable policies as well as providing primitives that help the community establish and adhere to the best practices of both K8s and its extensions.

Reviewer: Try something like:
```markdown
For Kubernetes v1.28, [ValidatingAdmissionPolicies](/docs/reference/access-authn-authz/validating-admission-policy/)
are now beta, and have gained some improvements:

- foo
- bar

To use ValidatingAdmissionPolicies, you need to enable the `admissionregistration.k8s.io/v1beta1` API group
in your cluster's control plane.
```

## CRD validation ratcheting
    

> This Allows custom resources to fail validation if the patch did not alter any of the invalid fields.  The ability to shift left validation logic from controllers to the front end is a long-term goal for improving the useability of the Kubernetes project.

Reviewer 14: ```suggestion
This allows custom resources to fail validation if the patch did not alter any of the invalid fields.  The ability to shift left validation logic from controllers to the front end is a long-term goal for improving the useability of the Kubernetes project.
```

> This Allows custom resources to fail validation if the patch did not alter any of the invalid fields.  The ability to shift left validation logic from controllers to the front end is a long-term goal for improving the useability of the Kubernetes project.

Reviewer 11: ```suggestion
This allows custom resources to fail validation if the patch did not alter any of the invalid fields.  The ability to shift left validation logic from controllers to the front end is a long-term goal for improving the useability of the Kubernetes project.
```

> This Allows custom resources to fail validation if the patch did not alter any of the invalid fields.  The ability to shift left validation logic from controllers to the front end is a long-term goal for improving the useability of the Kubernetes project.

Reviewer 15: > The ability to shift left validation logic from controllers to the front end is a long-term goal for improving the useability of the Kubernetes project.

For CRD Validation Ratcheting specifically, I'm not sure this is accurate statement in isolation. CRD Validation Ratcheting was added in part to support the Declarative Validation effort which does have an effect of shifting validation left, ratcheting errors for CRDs does not itself do this.

I'd say for K8S users the benefits are:

1. Improved UX knowing that previously applied resources will remain applied
2. CRDs can change over time to allow for most validations to be made more strict without requiring a new version and conversion webhook.

Author: Hi we have decided not to add this feature to the blog in the end after discussion, we didn't find a way to present it to readers without confusing them. Thanks for your help we appreciate it.

    
The Kubernetes yearly support period already makes annual upgrades possible. Users can upgrade to the latest patch versions to pick up security fixes and do 3 sequential minor version upgrades once a year to "catch up" to the latest supported minor version.

However, since the tested/supported skew between nodes and control planes is currently limited to 2 versions, a 3-version upgrade would have to update nodes twice to stay within the supported skew.

## Beta support for enabling swap space on Linux
    
This adds swap support to nodes in a controlled, predictable manner so that Kubernetes users can perform testing and provide data to continue building cluster capabilities on top of swap.

There are two distinct types of users for swap, who may overlap:

- Node administrators, who may want swap available for node-level performance tuning and stability/reducing noisy neighbor issues.
    

> - Application developers, who have written applications that would benefit from using swap memory. There are hence a number of possible ways that one could envision swap use on a node.

Reviewer 12: This seems unnecessary: "There are hence a number of possible ways that one could envision swap use on a node."

Reviewer: Based on https://github.com/kubernetes/website/pull/42381/files#r1293746761 I've made that into a GUI-committable fix
```suggestion
- Application developers, who have written applications that would benefit from using swap memory.
```

The mixed version proxy finds requests that the local API server doesn't recognize but another API server
inside the control plan is able to support. Having found a suitable peer, the aggregation layer proxies
the request to a compatible API server; this is transparent from the client's perspective.
    
When an upgrade or downgrade is performed on a cluster, for some period of time the API servers
within the control plane may be at differing versions; when that happens, different subsets of the
API servers are able to serve different sets of built-in resources (different groups, versions, and resources
are all possible). This new alpha mechanism lets you hide that skew from clients.
  
## Source code reorganization for control plane components

Kubernetes contributors have begun to reorganize the code for the kube-apiserver to build on a new staging repository that consumes [k/apiserver](https://github.com/kubernetes/apiserver) but has a bigger, carefully chosen subset of the functionality of kube-apiserver such that it is reusable.

This is a gradual reorganization; eventually there will be a new

> [k/generic-controlplane](https://github.com/kubernetes/generic-controlplane) Git repository

Reviewer 7: We should remove this link if the repo doesn't exist yet

Author: Thanks done.

For init containers, you can either omit the `restartPolicy` field, or set it to `Always`. Omitting the field
means that you want a true init container that runs to completion before application startup.

Sidecar containers do not block Pod completion: if all regular containers are complete, sidecar
containers in that Pod will be terminated.

During the sidecar startup stage, the restart behavior will be similar to init containers. If the Pod restartPolicy is Never, the sidecar container that failed during startup will NOT be restarted and the whole Pod will fail. If the Pod restartPolicy is Always or OnFailure, it will be restarted.

Once the sidecar container is started (postStart completed and startup probe succeeded), these containers will be restarted even when the Pod restartPolicy is Never or OnFailure. Furthermore, sidecar containers will be restarted even during Pod termination.

> 
> In order to minimize OOM kills of sidecar containers, the OOM adjustment for these containers will match or exceed the OOM score adjustment of regular containers in the Pod.
> 
> This also enables sidecar containers (those will not be allowed for other init containers):
> 
> - PostStart and PreStop lifecycle handlers for sidecar containers
> - All probes (startup, readiness, liveness)
>     
> Readiness probes of sidecars will contribute to determining the whole Pod readiness.
> 

Reviewer: ```suggestion

To learn more, read [API for sidecar containers](/docs/concepts/workloads/pods/init-containers/#api-for-sidecar-containers).

```

    
This changes the behavior of default storage class assignment to be retroactive for existing unbound persistent volume claims without any storage class assigned. This changes the existing Kubernetes behavior slightly, which is further described in the sections below.
             
## Pod replacement policy for Jobs (alpha) {#pod-replacement-policy}

Kubernetes 1.28 adds a new field for the Job API that allows you to specify if you want the control
plane to make new Pods as soon as the previous Pods begin termination (existing behavior),
 or only once the existing pods are fully terminated (new, optional behavior).
    
Many common machine learning frameworks, such as Tensorflow and JAX, require unique pods per index.
With the older behaviour, if a pod that belongs to an `Indexed` Job enters a terminating state (due to preemption, eviction or other external factors), a replacement pod is created but then immediately fails to start due
to the clash with the old pod that has not yet shut down.

Having a replacement Pod appear before the previous one fully terminates can also cause problems

> in clusters with scarce resources or with tight budgets. These resources can be difficult to obtain so 

Reviewer: You can make GitHub suggestions @mickeyboxell and I encourage that you do.

> in clusters with scarce resources or with tight budgets. These resources can be difficult to obtain so 

Author: Thanks done.

in the Job documentation.
  
## Job retry backoff limit, per index (alpha) {#job-per-index-retry-backoff}
    
This extends the Job API to support indexed jobs where the backoff limit is per index, and the Job can continue execution despite some of its indexes failing.
    
Currently, the indexes of an indexed job share a single backoff limit. When the job reaches this shared backoff limit, the job controller marks the entire job as failed, and the resources are cleaned up, including indexes that have yet to run to completion.

As a result, the existing implementation did not cover the situation where the workload is truly
[embarrassingly parallel](https://en.wikipedia.org/wiki/Embarrassingly_parallel): each index is
fully independent of other indexes.

For instance, if indexed jobs were used as the basis for a suite of long-running integration tests, then each test run would only be able to find a single test failure.

    
For more information, read [Handling Pod and container failures](/docs/concepts/workloads/controllers/job/#handling-pod-and-container-failures) in the Kubernetes documentation.
   
## CRI container and pod statistics without cAdvisor

> This encompasses two related pieces of work (changes to the kubelet's `/metrics/cadvisor` endpoint, and improvements to the replacement _summary_ API).

Reviewer 12: Nit: no need for comma.
```suggestion
This encompasses two related pieces of work (changes to the kubelet's `/metrics/cadvisor` endpoint and improvements to the replacement _summary_ API).
```

* [`kubectl events`](https://github.com/kubernetes/enhancements/issues/1440)
* [Retroactive default StorageClass assignment](https://github.com/kubernetes/enhancements/issues/3333)
* [Non-graceful node shutdown](https://github.com/kubernetes/enhancements/issues/2268)
* [Support 3rd party device monitoring plugins](https://github.com/kubernetes/enhancements/issues/606)
* [Auth API to get self-user attributes](https://github.com/kubernetes/enhancements/issues/3325)
* [Proxy Terminating Endpoints](https://github.com/kubernetes/enhancements/issues/1669)
* [Expanded DNS Configuration](https://github.com/kubernetes/enhancements/issues/2595)
* [Cleaning up IPTables Chain Ownership](https://github.com/kubernetes/enhancements/issues/3178)
* [Minimizing iptables-restore input size](https://github.com/kubernetes/enhancements/issues/3453)
* [Graduate the kubelet pod resources endpoint to GA](https://github.com/kubernetes/enhancements/issues/3743)
* [Extend podresources API to report allocatable resources](https://github.com/kubernetes/enhancements/issues/2403)
* [Move EndpointSlice Reconciler into Staging](https://github.com/kubernetes/enhancements/issues/3685)

### Deprecations And Removals

> Removals:

Author: Thanks done

* [Ceph FS in-tree plugin](https://github.com/kubernetes/kubernetes/pull/118143)

## Release Notes

The complete details of the Kubernetes v1.28 release are available in our [release notes](https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.28.md).

## Availability

Kubernetes v1.28 is available for download on [GitHub](https://github.com/kubernetes/kubernetes/releases/tag/v1.28.0). To get started with Kubernetes, you can run local Kubernetes clusters using [minikube](https://minikube.sigs.k8s.io/docs/), [kind](https://kind.sigs.k8s.io/), etc. You can also easily install v1.28 using [kubeadm](https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/create-cluster-kubeadm/).
 
## Release Team

Kubernetes is only possible with the support, commitment, and hard work of its community. Each release team is made up of dedicated community volunteers who work together to build the many pieces that make up the Kubernetes releases you rely on. This requires the specialized skills of people from all corners of our community, from the code itself to its documentation and project management.

> We would like to thank the entire release team for the hours spent hard at work to ensure we deliver a solid Kubernetes v1.28 release for our community.

Reviewer 16: ```suggestion
We extend our gratitude to the entire Release Teams, whose dedication and tireless efforts have been instrumental in crafting the Kubernetes v1.28 release for our valued community.
```

> Special thanks to our Release Lead, Grace Nguyen, for guiding us through a smooth and successful release cycle and facilitating the communication and collaboration between the different Release Teams taking advantage of her experience as a Shadow and Lead in these teams.

Author: Thanks i have dropped to lowercase

---
layout: blog
title: "Kubernetes v1.28: <name-here>"
date: 2023-08-02
slug: kubernetes-v1-28-release
---

**Authors**: [Kubernetes v1.28 Release Team](https://github.com/kubernetes/sig-release/blob/master/releases/release-1.28/release-team.md)

Announcing the release of Kubernetes v1.28 <name-here>, the second release of 2023!

This release consists of 46 enhancements. 20 of those enhancements are entering Alpha, 14 are graduating to Beta, and 12 are graduating to Stable.

## Release Theme And Logo

> **Kubernetes v1.28: *<name-here>*

Reviewer: Check formatting, eg:
```diff
-**Kubernetes v1.28: *<name-here>*
+**Kubernetes v1.28: <name-here>**
```

For Kubernetes v1.28, you can enable the _mixed version proxy_ within the API server's aggregation layer.
The mixed version proxy finds requests that the local API server doesn't recognize but another API server
inside the control plan is able to support. Having found a suitable peer, the aggregation layer proxies
the request to a compatible API server; this is transparent from the client's perspective.
    
When an upgrade or downgrade is performed on a cluster, for some period of time the API servers
within the control plane may be at differing versions; when that happens, different subsets of the
API servers are able to serve different sets of built-in resources (different groups, versions, and resources
are all possible). This new alpha mechanism lets you hide that skew from clients.

## CRD validation ratcheting
    
This Allows CRs to fail validation if the patch did not alter any of the invalid fields.  The ability to shift left validation logic from controllers to the front end is a long-term goal for improving the useability of the Kubernetes project.
    
## Source code reorganization for control plane components

> This factors the kube-apiserver to build on a new staging repository that consumes k/apiserver but has a bigger, carefully chosen subset of the functionality of kube-apiserver such that it is reusable.

Reviewer: ```suggestion
Kubernetes contributors have begun to reorganize the code for the kube-apiserver to build on a new staging repository that consumes [k/apiserver](https://github.com/kubernetes/apiserver) but has a bigger, carefully chosen subset of the functionality of kube-apiserver such that it is reusable.
```

> The factoring will be progressive: we will start with a new repo that adds
> nothing to k/apiserver, and then progressively move generic functionality from
> kube-apiserver to the new repository. The new repo will be named
> k/generic-controlplane.

Reviewer: ```suggestion
This is a gradual reorganization; eventually there will be a new
[k/generic-controlplane](https://github.com/kubernetes/generic-controlplane) Git repository
with generic functionality abstracted from Kubernetes' API server.
```

Once the sidecar container is started (postStart completed and startup probe succeeded), these containers will be restarted even when the Pod restartPolicy is Never or OnFailure. Furthermore, sidecar containers will be restarted even during Pod termination.

In order to minimize OOM kills of sidecar containers, the OOM adjustment for these containers will match or exceed the OOM score adjustment of regular containers in the Pod.

This also enables sidecar containers (those will not be allowed for other init containers):

- PostStart and PreStop lifecycle handlers for sidecar containers
- All probes (startup, readiness, liveness)
    
Readiness probes of sidecars will contribute to determining the whole Pod readiness.
    
## Beta support for enabling swap space on Linux
    
This adds swap support to nodes in a controlled, predictable manner so that Kubernetes users can perform testing and provide data to continue building cluster capabilities on top of swap.

There are two distinct types of users for swap, who may overlap:

- Node administrators, who may want swap available for node-level performance tuning and stability/reducing noisy neighbor issues.
    

> - Application developers, who have written applications that would benefit from using swap memory. There are hence a number of possible ways that one could envision swap use on a node.

Reviewer: I'd move this ahead of all the alpha features.

When a node is shutdown but not detected by Kubelet's Node Shutdown Manager, the pods that are part of a StatefulSet will be stuck in terminating status on the shutdown node and cannot move to a new running node. This is because Kubelet on the shutdown node is not available to delete the pods so the StatefulSet cannot create a new pod with the same name. If there are volumes used by the pods, the VolumeAttachments will not be deleted from the original shutdown node so the volumes used by these pods cannot be attached to a new running node. As a result, the application running on the StatefulSet cannot function properly. If the original shutdown node comes up, the pods will be deleted by Kubelet and new pods will be created on a different running node. If the original shutdown node does not come up, these pods will be stuck in terminating status on the shutdown node forever.

This handles node shutdown cases that are not detected by Kubelet's Node Shutdown Manager. The pods will be forcefully deleted in this case, triggering the deletion of the VolumeAttachments, and new pods will be created on a new running node so that application can continue to function.

> ## Automatic, retroactive assignment of a default StorageClass

Reviewer: The section here is too long really. Can we summarize or link to a previous release announcement?

When users want to change the default SC parameters, they must delete the SC and re-create it, Kubernetes API does not allow change in the SC. So there is no default SC for some time and the second case above applies here too. Re-creating the storage class to change parameters can be useful in cases where there is a quota set for the SC, and since the quota is coupled with the SC name users can not use Option 1 because the second SC would have a different name and so the existing quota would not apply to it.

Defined ordering during cluster installation. Kubernetes cluster installation tools must be currently smart enough to create a default SC before starting anything that may create PVCs that need it. If such a tool supports multiple cloud providers, storage backends, and add-ons that require storage (such as an image registry), it may be quite complicated to do the ordering right.
          

> ## Changes to supported skew between control plane and node versions

Reviewer: I'd list this ahead of any or beta feature.

Reviewer: The section is quite long. Is there documentation we can link to so the section in the release announcement stays small?

Author: Thanks, I have reached out for help to find the docs and have taken out some of the text. Will add link once i get it

    
This enables testing and expanding the supported skew between core node and control plane components by one version from n-2 to n-3, so that node components (kubelet and kube-proxy) for the oldest supported minor version work with control plane components (kube-apiserver, kube-scheduler, kube-controller-manager, cloud-controller-manager) for the newest supported minor version.
    
The Kubernetes yearly support period already makes annual upgrades possible. Users can upgrade to the latest patch versions to pick up security fixes and do 3 sequential minor version upgrades once a year to "catch up" to the latest supported minor version.

However, since the tested/supported skew between nodes and control planes is currently limited to 2 versions, a 3-version upgrade would have to update nodes twice to stay within the supported skew. For example, to upgrade from v1.40 to v1.43:

Begin: control plane and nodes on v1.40
Control plane upgrade: v1.40 → v1.41 → v1.42
Node upgrades: v1.40 → v1.42
Control plane upgrade: v1.42 → v1.43
Node upgrades: v1.42 → v1.43
    
Node upgrades are inherently more disruptive than control plane upgrades to workloads, for several reasons:

> Workloads can be designed to have no dependencies on the Kubernetes control plane, so Kubernetes control plane availability does not directly impact running pods

Reviewer: ```suggestion
Workloads can be designed to have no dependencies on the Kubernetes control plane, so Kubernetes control plane availability does not directly impact running pods.
```

There can be many more nodes (hundreds to thousands) than control plane members (typically 1 or 3).
    
Every time nodes are upgraded to a new minor version, every pod running on those nodes must be drained/rescheduled. This is true for immutable nodes and mutable/bare-metal nodes. If all nodes are being upgraded, this means every pod in the cluster will be replaced at least once. Patch updates of kubelet / kube-proxy components can be done in place, so it is possible to pick up security fixes and patch updates less disruptively.
    
Replacing or moving pods that are slow to stop or start or have significant data gravity takes significant time, so it is desirable to minimize how frequently that must be done.
If node/control plane skew support was expanded so the oldest node components work with the newest control plane components, the example upgrade path from v1.40 to v1.43 above could improve this:

Begin: control plane and nodes on v1.40
Control plane upgrade: v1.40 → v1.41 → v1.42 → v1.43
Node upgrades: v1.40 → v1.43
   

> ## Pod replacement policy for Jobs

Reviewer: ```suggestion
## Pod replacement policy for Jobs (alpha) {#pod-replacement-policy}
```

plane to make new Pods as soon as the previous Pods begin termination (existing behavior),
 or only once the existing pods are fully terminated (new, optional behavior).
    
Many common machine learning frameworks, such as Tensorflow and JAX, require unique pods per index.
With the older behaviour, if a pod that belongs to an `Indexed` Job enters a terminating state (due to preemption, eviction or other external factors), a replacement pod is created but then immediately fails to start due
to the clash with the old pod that has not yet shut down.

Having a replacement Pod appear before the previous one fully terminates can also cause problems
in clusters with scarce resources or with tight budgets. These resources can be difficult to obtain so 
can take a long time to find resources and they may only be able to find nodes once the existing pods
have been terminated. If cluster autoscaler is enabled, early creation of replacement Pods might produce undesired scale-ups.

To learn more, read [Delayed creation of replacement pods](/docs/concepts/workloads/controllers/job/#delayed-creation-of-replacement-pods)
in the Job documentation.
  

> ## Job retry backoff limit, per index

Reviewer: ```suggestion
## Job retry backoff limit, per index (alpha) {#job-per-index-retry-backoff}
```

    
This extends the Job API to support indexed jobs where the backoff limit is per index, and the Job can continue execution despite some of its indexes failing.
    
Currently, the indexes of an indexed job share a single backoff limit. When the job reaches this shared backoff limit, the job controller marks the entire job as failed, and the resources are cleaned up, including indexes that have yet to run to completion.

As a result, the existing implementation did not cover the situation where the workload is truly
[embarrassingly parallel](https://en.wikipedia.org/wiki/Embarrassingly_parallel): each index is
fully independent of other indexes.

For instance, if indexed jobs were used as the basis for a suite of long-running integration tests, then each test run would only be able to find a single test failure.

    
For more information, read [Handling Pod and container failures](/docs/concepts/workloads/controllers/job/#handling-pod-and-container-failures) in the Kubernetes documentation.
   

> ## cAdvisor-less, CRI-full Container And Pod Stats

Reviewer: ```suggestion
## CRI container and pod statistics without cAdvisor
```

This encompasses two related pieces of work (changes to the kubelet's `/metrics/cadvisor` endpoint, and improvements to the replacement _summary_ API).

There are two main APIs that consumers use to gather stats about running containers and pods: summary API and /metrics/cadvisor. The Kubelet is responsible for implementing the summary API, and cadvisor is responsible for fulfilling /metrics/cadvisor.

This aims to enhance CRI implementations to be able to fulfill all the stats needs of Kubernetes. At a high level, there are two pieces of this:

- Enhance the CRI API with enough metrics to be able to supplement the pod and container fields in the summary API directly from CRI.
    

> - Enhance the CRI implementations to broadcast the required metrics to fulfill the pod and container fields in the /metrics/cadvisor endpoint.

Reviewer: Rewrite this as if the feature is already released (it will have been). Also, use Markdown (eg backticks around URI paths).
